core/embed/prodtest/main.c

```diff
@@ -513,9 +603,446 @@ static void test_otp_write_device_variant(const char *args) {
+
+static bool set_metadata(uint16_t oid, const optiga_metadata *metadata) {
+  uint8_t serialized[258] = {0};
+  size_t size = 0;
+  optiga_result ret = optiga_serialize_metadata(metadata, serialized,
+                                                sizeof(serialized), &size);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_serialize_metadata error %d for OID 0x%04x.", ret,
+                oid);
+    return false;
+  }
+
+  optiga_set_data_object(oid, true, serialized, size);
+
+  ret =
+      optiga_get_data_object(oid, true, serialized, sizeof(serialized), &size);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_get_metadata error %d for OID 0x%04x.", ret, oid);
+    return false;
+  }
+
+  optiga_metadata metadata_stored = {0};
+  ret = optiga_parse_metadata(serialized, size, &metadata_stored);
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_parse_metadata error %d.", ret);
+    return false;
+  }
+
+  if (!optiga_compare_metadata(metadata, &metadata_stored)) {
+    vcp_println("ERROR optiga_compare_metadata failed.");
+    return false;
+  }
+
+  return true;
+}
+
+static bool pair_optiga(void) {
+  // The pairing key may already be written and locked. The success of the
+  // pairing procedure is determined by optiga_sec_chan_handshake(). Therefore
+  // it is OK for some of the intermediate operations to fail.
+
+  // Enable writing the pairing secret to OPTIGA.
+  optiga_metadata metadata = {0};
+  metadata.change = OPTIGA_ACCESS_ALWAYS;
+  metadata.execute = OPTIGA_ACCESS_ALWAYS;
+  metadata.data_type = TYPE_PTFBIND;
+  set_metadata(OID_KEY_PAIRING, &metadata);  // Ignore result.
+
+  // Generate pairing secret.
+  uint8_t secret[SECRET_OPTIGA_KEY_LEN] = {0};
+  optiga_result ret = optiga_get_random(secret, sizeof(secret));
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_get_random error %d,", ret);
+    return false;
+  }
+
+  // Store pairing secret.
+  ret = optiga_set_data_object(OID_KEY_PAIRING, false, secret, sizeof(secret));
+  /*
+   * TODO: Uncomment. Right now this code will render the device unusable with
+   * unofficial firmware. We need to be able to call AttestationDelete before
+   * this code is enabled.
+   *
+  if (OPTIGA_SUCCESS == ret) {
+    secret_erase();
+    secret_write_header();
+    secret_write(secret, SECRET_OPTIGA_KEY_OFFSET, SECRET_OPTIGA_KEY_LEN);
+  }
+
+  // Verify whether the secret was stored correctly in flash and OPTIGA.
+  memzero(secret, sizeof(secret));
+  if (secret_read(secret, SECRET_OPTIGA_KEY_OFFSET, SECRET_OPTIGA_KEY_LEN) !=
+      sectrue) {
+    vcp_println("ERROR Failed to read pairing secret.");
+    return false;
+  }
+  */
+
+  ret = optiga_sec_chan_handshake(secret, sizeof(secret));
+  memzero(secret, sizeof(secret));
+  if (OPTIGA_SUCCESS != ret) {
+    vcp_println("ERROR optiga_sec_chan_handshake error %d.", ret);
+    return false;
+  }
+
+  return true;
+}
+
+static void optiga_lock(void) {
+  if (!pair_optiga()) {
```

Comment:
given that this is the only place where `pair_optiga()` is called, will it work together with `drop_privileges()` which happens before it?

or should we add `pair_optiga()` to the main flow after Optiga initialization?

Reply:
The function needs to be able to call `secret_erase()`, `secret_write_header()`, `secret_write()` and `secret_read()`. Is that possible even after `drop_privileges()`? @TychoVrahe

Assuming that the answer is yes, I think that in principle `pair_optiga()` can be called in either place. But I wonder whether it makes sense to call it automatically at every startup.

Reply:
Seems like MPU is still set to allow the secret reading/writing after privileges get dropped. Do we wanna prevent this?

Reply:
Ideally, yes. The main idea about all this is that if there is a vulnerability in prodtest, it should not be possible to exfiltrate the secret through it.

Reply:
Done in f92db56fc8319aa573536e39396597e0e99ebd46. But I am not sure how to handle the error state. Should we make sure to drop privileges? How about going to the main loop and responding with `ERROR failed to pair with Optiga` for every command?

Reply:
this commit https://github.com/trezor/trezor-firmware/commit/21047b29370fdadd64f24422d9fa3d14b5529a9f disables access to the secret after privileges are dropped. untested though.

Reply:
But what if the pairing fails? It seems better to return some sort of error over the VCP rather than shutting down. Any opinions about what I proposed about going to the main loop?

Reply:
We can do the VCP response to subsequent commands, or just print an error message on display and shutdown. The former allows non-optiga related part of the test to be run, with some added complexity to the code. Not sure whether the ability to run the test further has any value for the manufacturer

Reply:
I just implemented it a moment ago: https://github.com/trezor/trezor-firmware/pull/3185/commits/0382bf6dcfffd08600362e5fe40a95bf3e190f62